packages/flutter/lib/src/foundation/persistent_hash_map.dart

```diff
@@ -0,0 +1,340 @@
+  /// mapping.
+  PersistentHashMap<K, V> copyWith(K key, V value) {
+    final _TrieNode newroot =
+        (_root ?? _CompressedNode.empty).copyWith(0, key, key.hashCode, value);
+    if (newroot == _root) {
+      return this;
+    }
+    return PersistentHashMap<K, V>._(newroot);
+  }
+
+  /// Returns value associated with the given [key] or `null` if [key]
+  /// is not in the map.
+  V? lookup(K key) {
+    return _root != null ? _root!.lookup(0, key, key.hashCode) as V? : null;
+  }
+}
+
+/// Base class for nodes in a hash trie.
+abstract class _TrieNode {
+  static const int hashBitsPerLevel = 5;
+  static const int hashBitsPerLevelMask = (1 << hashBitsPerLevel) - 1;
+
+  @pragma('vm:prefer-inline')
+  static int trieIndex(int hash, int bitIndex) {
+    return (hash >>> bitIndex) & hashBitsPerLevelMask;
+  }
+
+  _TrieNode copyWith(int bitIndex, Object key, int keyHash, Object? value);
+
+  Object? lookup(int bitIndex, Object key, int keyHash);
+}
+
+/// A full (uncompressed) node in the trie.
+///
+/// It contains an array with `1<<_hashBitsPerLevel` elements which
+/// are references to deeper nodes.
+class _FullNode extends _TrieNode {
+  _FullNode(this.descendants);
+
+  static const int numElements = 1 << _TrieNode.hashBitsPerLevel;
+
+  final List<Object?> descendants;
+
+  @override
+  _TrieNode copyWith(int bitIndex, Object key, int keyHash, Object? value) {
+    final int index = _TrieNode.trieIndex(keyHash, bitIndex);
+    final _TrieNode node =
+        (descendants[index] as _TrieNode?) ?? _CompressedNode.empty;
+    final _TrieNode newNode = node.copyWith(
+        bitIndex + _TrieNode.hashBitsPerLevel, key, keyHash, value);
+    return identical(newNode, node)
+        ? this
+        : _FullNode(_copy(descendants)..[index] = newNode);
+  }
+
+  @override
+  Object? lookup(int bitIndex, Object key, int keyHash) {
+    final int index = _TrieNode.trieIndex(keyHash, bitIndex);
+    final _TrieNode? node = descendants[index] as _TrieNode?;
+    return node?.lookup(bitIndex + _TrieNode.hashBitsPerLevel, key, keyHash);
+  }
+}
+
+/// Compressed node in the trie.
+///
+/// Instead of storing the full array of outgoing edges this node uses a
+/// compressed representation:
+///
+///   * [_CompressedNode.occupied] has a bit set for indices which are occupied.
+///   * furthermore, each occupied index can either be a `(key, value)` pair
+///     representing an actual key/value mapping or a `(null, trieNode)` pair
+///     representing a descendant node.
+///
+class _CompressedNode extends _TrieNode {
+  _CompressedNode(this.occupiedIndices, this.keyValuePairs);
+  _CompressedNode._empty() : this(0, _emptyArray);
+
+  factory _CompressedNode.single(int bitIndex, int keyHash, _TrieNode node) {
+    final int bit = 1 << _TrieNode.trieIndex(keyHash, bitIndex);
+    // A single (null, node) pair.
+    final List<Object?> keyValuePairs = List<Object?>.filled(2, null)
+      ..[1] = node;
```

Comment:
Also: why not have two arrays, one for keys and one for values?

Reply:
It is better for locality of data and memory usage. 

```
final growable = [a, b, c, d];
final fixed = List.filled(4, null);
```

`growable` is two objects (growable array object pointing to a fixed array object which acts as a backing storage). `fixed` is a single object (fixed length array object). Thus allocating `growable` requires more allocations (and more memory) and accessing elements in `growable` requires traversing an indirection.

The same applies to packing key-value pairs together. Improves locality and decreases amount of allocated objects also saves a bit of memory - due to padding and object header overhead. Remember that 1 element array takes 4 words, 2 element array takes 6, 3 element 6, 4 element 8, 5 element 8, 6 element 10. So two 1 element arrays take 8 while 2 element array would take 6) and so on. 

That being said: this likely can be tweaked for `dart2js` where underlying platform does not have efficient representation for fixed length arrays and it might be actually more performant to use growable arrays. (Fixed arrays actually require an aritificial marker property in `dart2js` if I remember correctly so are slower to create). I will do some measurements.